### PR TITLE
Add support for static linking of the L0 adapter

### PR DESCRIPTION
**Pull request by pbalcer**
This patch adds a build-time option that builds the L0 adapter
as a static library and then embeds its in the loader.
The mechanim itself is fairly generic and, when a need arises,
the same can be done for other adapters.

### CODE REVIEW

File: source/adapters/level_zero/image.cpp
Comment (isaacault) [inline source/adapters/level_zero/image.cpp:759]: Personally, I don't like having namespace twice with this gap in the middle. I'd prefer if the copy function were moved below to allow for one contiguous namespace declaration in the file.

Comment (pbalcer) [reply source/adapters/level_zero/image.cpp:759]: I wanted to avoid moving too much stuff, but it seems unavoidable. I'll just add the namespace for everything.

**Review by EwanC — APPROVED** [review]

**Review by isaacault — APPROVED** [review]
Bindless Images changes LGTM.

**Comment by pbalcer** [comment]
@nrspruit can you please take a look?

**Comment by pbalcer** [comment]
@oneapi-src/unified-runtime-level-zero-write @oneapi-src/unified-runtime-maintain This is a nightmare to rebase (I'm not doing it again until I have a full set of approvals :-)), I'd appreciate a prompt review.

**Review by kbenzie — APPROVED** [review]

**Review by nrspruit — APPROVED** [review]
LGTM

**Comment by pbalcer** [comment]
intel/llvm build is failing with `pi_level_zero.cpp:(.text.piextPlatformGetNativeHandle+0x32): undefined reference to `urPlatformGetNativeHandle'. I'm not sure whether changing the PI layer right now is worth it, so I'm going to wait with this patch until that layer is removed.

File: .github/workflows/build-hw-reusable.yml
Comment (kswiecicki) [inline .github/workflows/build-hw-reusable.yml:40]: It seems that this adds L0 HW job to `OpenCL, HIP, CUDA ...` runners.

Comment (pbalcer) [reply .github/workflows/build-hw-reusable.yml]: done.

**Pull request merged**
